Automattic/newspack-popups
# feat: only show prompts for the best-priority segment

### All Submissions:






### Changes proposed in this Pull Request:



An experiment. Changes the concept of segment priority so that instead of showing prompts belonging to all of the reader's matching segments, we _only_ show the prompts that match the reader's highest-priority segment. This applies to prompts of all types (overlay, above-header, and inline), but we can choose to ignore this for inlines if we want.

With this paradigm, prompts with the default/no segment (here renamed as "Everyone" to more accurately convey its behavior along with https://github.com/Automattic/newspack-plugin/pull/853) take on increased importance. While the "Everyone" segment is still always prioritized last when it comes to conflict resolution for overlays and above-headers—so that any overlay or above-header prompt assigned to a reader-matching segment will be displayed instead of those assigned to "Everyone"—any _non-conflicting_ prompt assigned to the "Everyone" segment will be displayed to all readers, regardless of segment.

This also somewhat simplifies the previewing logic. Now, since there's no cascade of matching segments, when previewing a segment, we only need to show the prompts belonging to that segment plus any prompts belonging to "Everyone".

~Adding the "On Hold" status because we're not sure we want to go with this approach, but we want to test it to see how it "feels" from the publisher's perspective. Note that unit tests are failing now; if we decide we want to go with this new approach, I'll update the tests to pass.~ I updated the tests with the new functions, so now they're all passing once again. We could potentially add more tests for the non-cascading functionality, but I'm open to doing ideas.

Closes #445.

### How to test the changes in this Pull Request:

1. On this branch, create several overlay, above-header and inline prompts belonging to several different segments with varying overlapping criteria, and at least one prompt belonging to the "Everyone" segment. Examples of segments which can easily overlap:
  - Segment A is New Readers, or max articles read = 10. This should be set to the highest priority.
  - Segment B is Non-subscribers, with Newsletter set to non-subscribers only. This should be set to the middle priority.
  - Segment C is Non-donors, with Donor set to non-donors only. This should be set to the lowest priority.
2. In the Campaigns tab, preview each segment. Confirm that you see only published and unpublished prompts belonging to that segment, plus any prompts belonging to "Everyone", and no others.
  - Also try previewing "Everyone" and confirm that you only see prompts belonging to "Everyone".
  - If there's more than one overlay or above-header with the same segment, you'll still only see the most recently created one (this behavior is unchanged).
3. In a fresh incognito session, view an article on the site. In a fresh session, you should match all three of the example segments above. Confirm that because Segment A is the highest-priority segment, you only see the prompts belonging to that segment, plus any prompts belonging to the "Everyone" segment.
4. In the admin, reorder the segments so that Segment B is now the highest-priority segment. View an article in a new incognito session and this time confirm that you only see prompts belonging to Segment B, plus any prompts belonging to the "Everyone" segment.
5. Reorder segments so that Segment C is now the highest-priority segment. View an article in a new incognito session and this time confirm that you only see prompts belonging to Segment C, plus any prompts belonging to the "Everyone" segment.
6. Remove all prompts from Segment C, keeping it as the highest-priority segment. View the site in a fresh incognito session and confirm that you now see only prompts belonging to the "Everyone" segment.
7. Confirm that the conflict resolution aspect of segment priority still works: publish two overlays and above headers and assign to different segments. When viewing in a fresh incognito session, you should now only see overlays and above-header prompts if they belong to the highest-priority segment that you match as a reader.

### Other information:
